File: frontend/packages/shared/src/schemas.ts
import { z } from "zod";
import { ACTIONS } from "./actions";
import { TONE_PRESETS } from "./tones";
import { MODEL_IDS } from "./models";
import { LANGUAGE_IDS, SOURCE_LANGUAGE_IDS } from "./languages";
import { LIMITS } from "./constants";
import { ERROR_CODES } from "./errors";

// ---------------------------------------------------------------------------
// /api/v1/complete  (Edge, streaming SSE)
// ---------------------------------------------------------------------------

// The page-extracted context. Each field is optional because adapters extract
// what they can; the prompt builder copes with missing pieces.
export const ContextSchema = z
  .object({
    site: z.string().min(1).max(120).optional(),
    pageTitle: z.string().max(300).optional(),
    pageUrl: z.string().url().max(2048).optional(),

    // For email/threaded conversations: the thread the user is replying to.
    thread: z
      .array(
        z.object({
          author: z.string().max(200).optional(),
          // Stripped HTML; bounded by MAX_CONTEXT_CHARS at the field level.
          text: z.string().max(LIMITS.MAX_CONTEXT_CHARS),
          timestamp: z.string().max(64).optional(),
        }),
      )
      .max(40)
      .optional(),

    // For posts/comments: a single piece of content to respond to.
    post: z
      .object({
        author: z.string().max(200).optional(),
        text: z.string().max(LIMITS.MAX_CONTEXT_CHARS),
      })
      .optional(),

    // For grammar/rewrite: what the user has typed so far.
    draft: z.string().max(LIMITS.MAX_DRAFT_CHARS).optional(),

    // Misc page context (subject line, channel name, etc.). Free-form bag, but
    // the server caps each value before forwarding.
    meta: z.record(z.string().max(500)).optional(),
  })
  .strict();

export type RequestContext = z.infer<typeof ContextSchema>;

// Optional personalization. The extension stores this locally (in
// chrome.storage.local) and attaches it to the request — there is no
// server-side account or database. The backend uses it only to shape the
// prompt; it is never persisted.
export const RequestProfileSchema = z
  .object({
    displayName: z.string().max(120).optional(),
    aboutMe: z.string().max(2000).optional(),
  })
  .strict();

export type RequestProfile = z.infer<typeof RequestProfileSchema>;

export const CompleteRequestSchema = z
  .object({
    action: z.enum(ACTIONS),
    context: ContextSchema,
    tone: z.enum(TONE_PRESETS).optional(),
    instruction: z.string().max(LIMITS.MAX_INSTRUCTION_CHARS).optional(),
    model: z.enum(MODEL_IDS).optional(),
    // Language controls (see ./languages).
    //   sourceLanguage — the language of the input. "auto" (or omitted)
    //     lets the model detect it; a concrete id is treated as a hint.
    //   targetLanguage — the desired output language. Required for the
    //     "translate" action; optional elsewhere (omitted = match source).
    //   bilingual — produce the response in BOTH the source and target
    //     languages (only meaningful for "reply" / "rewrite").
    sourceLanguage: z.enum(SOURCE_LANGUAGE_IDS).optional(),
    targetLanguage: z.enum(LANGUAGE_IDS).optional(),
    bilingual: z.boolean().optional(),
    // Client-supplied personalization, sourced from local extension storage.
    profile: RequestProfileSchema.optional(),
    // Client-generated request id, echoed in metadata logs.
    clientRequestId: z.string().uuid().optional(),
  })
  .strict()
  .refine(
    (val) => {
      const hasDraft = !!(val.context.draft && val.context.draft.length > 0);
      const hasInstruction = !!(val.instruction && val.instruction.trim().length > 0);
      const hasPageContext = !!(val.context.thread?.length || val.context.post);

      if (val.action === "grammar") {
        // Grammar fixes a piece of text — it's meaningless without a draft.
        return hasDraft;
      }
      if (val.action === "translate") {
        // Translate needs text to translate and a language to translate into.
        return (hasDraft || hasPageContext) && !!val.targetLanguage;
      }
      if (val.action === "rewrite") {
        // Rewrite is broader than its name: it's also the "compose from
        // description" path. Accept any of:
        //   - a draft to transform (classic rewrite),
        //   - an instruction telling us what to write (compose),
        //   - page context to draw from (e.g., "summarize this thread").
        return hasDraft || hasInstruction || hasPageContext;
      }
      // reply
      return hasPageContext;
    },
    {
      message:
        "Action requires content to work with: 'reply' needs thread/post, " +
        "'grammar' needs a draft, 'translate' needs text plus a target " +
        "language, 'rewrite' needs at least one of draft, instruction, or " +
        "page context.",
      path: ["context"],
    },
  );

export type CompleteRequest = z.infer<typeof CompleteRequestSchema>;

// SSE payloads — what we emit to the extension on the wire.
export const SseTokenPayloadSchema = z.object({
  delta: z.string(),
});
export type SseTokenPayload = z.infer<typeof SseTokenPayloadSchema>;

export const SseUsagePayloadSchema = z.object({
  promptTokens: z.number().int().nonnegative(),
  completionTokens: z.number().int().nonnegative(),
  totalTokens: z.number().int().nonnegative(),
  model: z.string(),
});
export type SseUsagePayload = z.infer<typeof SseUsagePayloadSchema>;

export const SseErrorPayloadSchema = z.object({
  code: z.nativeEnum(ERROR_CODES),
  message: z.string(),
  retryable: z.boolean(),
});
export type SseErrorPayload = z.infer<typeof SseErrorPayloadSchema>;

// ---------------------------------------------------------------------------
// Local extension settings (stored in chrome.storage.local — never sent to
// the backend except for the `profile` subset attached to /complete).
// ---------------------------------------------------------------------------

export const LocalSettingsSchema = z
  .object({
    displayName: z.string().max(120),
    aboutMe: z.string().max(2000),
    defaultTone: z.enum(TONE_PRESETS),
    defaultModel: z.enum(MODEL_IDS),
    // Multilingual preferences. `workingLanguage` is the agent's preferred
    // language for drafting and the second language in bilingual replies;
    // `frequentLanguages` are surfaced first in the popover's language
    // pickers so common target languages are one click away.
    workingLanguage: z.enum(LANGUAGE_IDS),
    frequentLanguages: z.array(z.enum(LANGUAGE_IDS)).max(20),
    siteAllowlist: z.array(z.string().min(1).max(200)).max(200),
    siteBlocklist: z.array(z.string().min(1).max(200)).max(200),
    // Backend the extension talks to. Defaults to the build-time
    // VITE_BACKEND_URL but is user-configurable at runtime, so the
    // extension can be pointed at any compatible backend (see
    // docs/how-to/use-your-own-backend.md). No trailing slash.
    backendUrl: z.string().url(),
    // Optional credential for the user's own backend, sent as
    // `Authorization: Bearer <apiKey>`. Empty = no header (the default
    // Inkwell backend needs none).
    apiKey: z.string().max(500),
  })
  .strict();

export type LocalSettings = z.infer<typeof LocalSettingsSchema>;

// ---------------------------------------------------------------------------
// /api/v1/ocr  (Node, JSON in/out, non-streaming)
//
// The image is sent as base64 so the wire payload is JSON-only (matches the
// /complete contract). MIME type tells the vision model how to decode it.
// The endpoint is anonymous and rate-limited by client IP, like /complete.
// ---------------------------------------------------------------------------

// Allowed image MIME types. Whitelisted rather than open-ended so a
// malicious client can't smuggle SVG (XSS surface in some renderers) or
// arbitrary binary blobs through the endpoint.
export const OCR_MIME_TYPES = ["image/png", "image/jpeg", "image/webp", "image/gif"] as const;
export type OcrMimeType = (typeof OCR_MIME_TYPES)[number];

export const OcrRequestSchema = z
  .object({
    /** Standard base64 (no `data:` prefix), max ~10 MB encoded. */
    imageBase64: z
      .string()
      .min(64, "Image data is too small to be a real image.")
      .max(
        Math.ceil((LIMITS.MAX_OCR_IMAGE_BYTES * 4) / 3) + 32,
        "Image data exceeds the OCR size limit.",
      )
      .regex(/^[A-Za-z0-9+/=\s]+$/, "Image data must be base64."),
    mimeType: z.enum(OCR_MIME_TYPES),
  })
  .strict();
export type OcrRequest = z.infer<typeof OcrRequestSchema>;

export const OcrResponseSchema = z
  .object({
    text: z.string(),
    /** Model id that produced the response; useful for debugging tier
     *  drift between deployments. Absent in mock responses. */
    model: z.string().optional(),
  })
  .strict();
export type OcrResponse = z.infer<typeof OcrResponseSchema>;

// ---------------------------------------------------------------------------
// /api/v1/health
// ---------------------------------------------------------------------------

export const HealthSchema = z
  .object({
    ok: z.literal(true),
    version: z.string(),
    runtime: z.string(),
    timestamp: z.string(),
  })
  .strict();

export type Health = z.infer<typeof HealthSchema>;

File: backend/src/inkwell_backend/domain/schemas.py
"""Pydantic v2 DTOs.

Mirrors the zod schemas in `@inkwell/shared/schemas`. Every request and
response that crosses the HTTP boundary has a Pydantic model here, so:

* the OpenAPI schema FastAPI generates is automatically accurate;
* misshaped requests fail at the boundary with a 400, never reach
  downstream code;
* the same DTOs are reused inside the pipelines for type-safe access.

Constraints:

* `model_config = ConfigDict(extra="forbid")` mirrors zod's ``.strict()``
  so unknown keys raise validation errors instead of being silently
  accepted (defence-in-depth against the client smuggling fields).
* Cross-field rules use ``@model_validator(mode="after")`` — Pydantic's
  equivalent of zod's ``.refine()``.
"""

from __future__ import annotations

from typing import Annotated, Literal, Self

from pydantic import (
    AnyHttpUrl,
    BaseModel,
    ConfigDict,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)

from .actions import Action
from .languages import AUTO_DETECT, LANGUAGE_IDS, SOURCE_LANGUAGE_IDS
from .limits import (
    MAX_CONTEXT_CHARS,
    MAX_DRAFT_CHARS,
    MAX_HISTORY_TURN_CHARS,
    MAX_HISTORY_TURNS,
    MAX_INSTRUCTION_CHARS,
    MAX_MODEL_ID_CHARS,
    MAX_OCR_IMAGE_BYTES,
)
from .models import MODEL_IDS
from .tones import TonePreset

# Bare-string constraint helpers — Pydantic equivalents of zod's
# ``.max(...)``. Keeping them named so the schema reads like prose.
SiteName = Annotated[str, StringConstraints(min_length=1, max_length=120)]
PageTitle = Annotated[str, StringConstraints(max_length=300)]
PageUrl = Annotated[AnyHttpUrl, Field(max_length=2048)]
Author = Annotated[str, StringConstraints(max_length=200)]
ContextText = Annotated[str, StringConstraints(max_length=MAX_CONTEXT_CHARS)]
DraftText = Annotated[str, StringConstraints(max_length=MAX_DRAFT_CHARS)]
InstructionText = Annotated[str, StringConstraints(max_length=MAX_INSTRUCTION_CHARS)]
MetaValue = Annotated[str, StringConstraints(max_length=500)]
Timestamp = Annotated[str, StringConstraints(max_length=64)]

# Allowed image MIME types — whitelisted so a malicious client can't smuggle
# SVG (XSS surface) or arbitrary binary blobs through the OCR endpoint.
OCR_MIME_TYPES: tuple[str, ...] = (
    "image/png",
    "image/jpeg",
    "image/webp",
    "image/gif",
)
OcrMimeType = Literal["image/png", "image/jpeg", "image/webp", "image/gif"]

# Language enum types — the literal unions are derived from the catalogs
# so the schemas stay in sync with the source of truth.
LanguageId = Annotated[str, Field(json_schema_extra={"enum": list(LANGUAGE_IDS)})]
SourceLanguage = Annotated[str, Field(json_schema_extra={"enum": list(SOURCE_LANGUAGE_IDS)})]
# Bounded charset + length so even the relaxed Portkey path (below) can't
# accept an unbounded or exotic string. Covers catalog ids (``gpt-4o``)
# and Portkey model-catalog slugs (``@bedrock-use1/us.anthropic.claude``).
# The ``enum`` hint documents the curated direct-OpenAI catalog in OpenAPI;
# the cross-field validator decides whether non-catalog ids are allowed.
ModelId = Annotated[
    str,
    StringConstraints(min_length=1, max_length=MAX_MODEL_ID_CHARS, pattern=r"^[A-Za-z0-9._:@/-]+$"),
    Field(json_schema_extra={"enum": list(MODEL_IDS)}),
]


class ThreadMessage(BaseModel):
    """A single message in a threaded conversation extracted from the page."""

    model_config = ConfigDict(extra="forbid")

    author: Author | None = None
    text: ContextText
    timestamp: Timestamp | None = None


class Post(BaseModel):
    """A single post / comment the user is responding to."""

    model_config = ConfigDict(extra="forbid")

    author: Author | None = None
    text: ContextText


class RequestContext(BaseModel):
    """Page context extracted by the in-page adapter.

    Each field is optional because adapters extract what they can; the
    prompt builder copes with missing pieces.
    """

    model_config = ConfigDict(extra="forbid")

    site: SiteName | None = None
    page_title: PageTitle | None = Field(default=None, alias="pageTitle")
    page_url: PageUrl | None = Field(default=None, alias="pageUrl")
    thread: list[ThreadMessage] | None = Field(default=None, max_length=40)
    post: Post | None = None
    draft: DraftText | None = None
    meta: dict[str, MetaValue] | None = None


HistoryTurnText = Annotated[str, StringConstraints(min_length=1, max_length=MAX_HISTORY_TURN_CHARS)]


class ConversationTurn(BaseModel):
    """One prior turn in a refinement conversation.

    The client replays earlier turns — its own task message(s) and the
    assistant's previous draft(s) — so a follow-up like "make it shorter"
    revises the existing output instead of starting from scratch. Roles
    mirror the chat convention; ``system`` is intentionally excluded so a
    client can never inject a competing system prompt.
    """

    model_config = ConfigDict(extra="forbid")

    role: Literal["user", "assistant"]
    text: HistoryTurnText


class RequestProfile(BaseModel):
    """Optional personalization the extension attaches from local storage."""

    model_config = ConfigDict(extra="forbid")

    display_name: Annotated[str, StringConstraints(max_length=120)] | None = Field(
        default=None,
        alias="displayName",
    )
    about_me: Annotated[str, StringConstraints(max_length=2000)] | None = Field(
        default=None,
        alias="aboutMe",
    )


class CompleteRequest(BaseModel):
    """Body of POST /api/v1/complete."""

    # `populate_by_name` lets pipeline code read snake_case attributes
    # while client requests stay camelCase on the wire.
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    action: Action
    context: RequestContext
    tone: TonePreset | None = None
    instruction: InstructionText | None = None
    model: ModelId | None = None
    source_language: SourceLanguage | None = Field(default=None, alias="sourceLanguage")
    target_language: LanguageId | None = Field(default=None, alias="targetLanguage")
    bilingual: bool | None = None
    # Prior turns for conversational refinement. Empty/omitted for the
    # first request; populated when the user iterates on a result.
    history: list[ConversationTurn] | None = Field(default=None, max_length=MAX_HISTORY_TURNS)
    profile: RequestProfile | None = None
    client_request_id: Annotated[str, Field(pattern=r"^[0-9a-fA-F-]{36}$")] | None = Field(
        default=None,
        alias="clientRequestId",
    )

    @field_validator("source_language")
    @classmethod
    def _source_language_in_catalog(cls, value: str | None) -> str | None:
        if value is None or value == AUTO_DETECT or value in LANGUAGE_IDS:
            return value
        raise ValueError(f"Unknown sourceLanguage: {value!r}")

    @field_validator("target_language")
    @classmethod
    def _target_language_in_catalog(cls, value: str | None) -> str | None:
        if value is None or value in LANGUAGE_IDS:
            return value
        raise ValueError(f"Unknown targetLanguage: {value!r}")

    @field_validator("model")
    @classmethod
    def _model_in_catalog(cls, value: str | None) -> str | None:
        if value is None or value in MODEL_IDS:
            return value
        # When Portkey is enabled the gateway owns model routing —
        # virtual keys and model-catalog slugs ("@integration/model")
        # reference upstreams the backend has no catalog for. Accept any
        # well-formed id (charset/length already bounded by ``ModelId``)
        # and let the gateway validate it. Direct OpenAI stays strict so
        # a typo fails fast at the boundary instead of burning a call.
        from ..settings import get_settings

        if get_settings().portkey_enabled:
            return value
        raise ValueError(f"Unknown model: {value!r}")

    @model_validator(mode="after")
    def _action_specific_content(self) -> Self:
        ctx = self.context
        has_draft = bool(ctx.draft and len(ctx.draft) > 0)
        has_instruction = bool(self.instruction and self.instruction.strip())
        has_page_context = bool((ctx.thread and len(ctx.thread) > 0) or ctx.post)

        if self.action is Action.GRAMMAR and not has_draft:
            raise ValueError("'grammar' action requires `context.draft` to be provided.")
        if self.action is Action.TRANSLATE:
            if not (has_draft or has_page_context):
                raise ValueError("'translate' action needs text in context.draft / thread / post.")
            if not self.target_language:
                raise ValueError("'translate' action requires `targetLanguage`.")
        if self.action is Action.REWRITE and not (has_draft or has_instruction or has_page_context):
            raise ValueError(
                "'rewrite' action needs at least one of context.draft, instruction, "
                "or context.thread/post."
            )
        if self.action is Action.REPLY and not has_page_context:
            raise ValueError("'reply' action requires context.thread or context.post.")
        if self.action in (Action.SUMMARIZE, Action.EXPLAIN) and not (
            has_draft or has_page_context
        ):
            raise ValueError(
                f"'{self.action.value}' action needs text in context.draft / thread / post."
            )
        return self


# ---------------------------------------------------------------------------
# SSE payloads — what the /complete stream emits.
# ---------------------------------------------------------------------------


class SseTokenPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")
    delta: str


class SseUsagePayload(BaseModel):
    # `populate_by_name` so the pipeline can construct this with
    # Python-natural snake_case kwargs; ``model_dump(by_alias=True)``
    # still emits camelCase on the wire.
    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    prompt_tokens: Annotated[int, Field(ge=0, alias="promptTokens")]
    completion_tokens: Annotated[int, Field(ge=0, alias="completionTokens")]
    total_tokens: Annotated[int, Field(ge=0, alias="totalTokens")]
    model: str


class SseErrorPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str
    message: str
    retryable: bool


# ---------------------------------------------------------------------------
# /api/v1/ocr
# ---------------------------------------------------------------------------


_MAX_OCR_B64_LEN = ((MAX_OCR_IMAGE_BYTES * 4) + 2) // 3 + 32


class OcrRequest(BaseModel):
    """Body of POST /api/v1/ocr."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    image_base64: Annotated[
        str,
        StringConstraints(
            min_length=64,
            max_length=_MAX_OCR_B64_LEN,
            pattern=r"^[A-Za-z0-9+/=\s]+$",
        ),
    ] = Field(alias="imageBase64")
    mime_type: OcrMimeType = Field(alias="mimeType")


class OcrResponse(BaseModel):
    """Body of a successful /api/v1/ocr 200."""

    model_config = ConfigDict(extra="forbid")

    text: str
    model: str | None = None


# ---------------------------------------------------------------------------
# /api/v1/health
# ---------------------------------------------------------------------------


class Health(BaseModel):
    """Body of a successful /api/v1/health 200."""

    model_config = ConfigDict(extra="forbid")

    ok: Literal[True]
    version: str
    runtime: str
    timestamp: str
